injected/src/url-change.js

```diff
@@ -0,0 +1,51 @@
+import { DDGProxy, DDGReflect, isBeingFramed } from './utils.js';
+import ContentFeature from './content-feature.js';
+
+const urlChangeListeners = new Set();
+/**
+ * Register a listener to be called when the URL changes.
+ * @param {function} listener
+ */
+export function registerForURLChanges(listener) {
+    if (urlChangeListeners.size === 0) {
+        listenForURLChanges();
+    }
+    urlChangeListeners.add(listener);
+}
+
+function handleURLChange() {
+    for (const listener of urlChangeListeners) {
+        listener();
+    }
+}
+
+function listenForURLChanges() {
+    const urlChangedInstance = new ContentFeature('urlChanged', {}, {});
+    if ('navigation' in globalThis && 'addEventListener' in globalThis.navigation) {
+        // if the browser supports the navigation API, we can use that to listen for URL changes
+        // Listening to navigatesuccess instead of navigate to ensure the navigation is committed.
+        globalThis.navigation.addEventListener('navigatesuccess', () => {
```

Comment:
@dbajpeyi the navigate event actually fired too early, I validated this works as expected. The API has a way to modify and deny navigations before they happen.